modules/fbs/includes/fbs.debt.inc

```diff
@@ -10,71 +10,89 @@
  *
  * @param object $account
  *   User to get debt listing for.
+ * @param bool $reset
+ *   If TRUE cache is reset (default: FALSE)
  *
  * @return array
  *   Debt information.
  */
-function fbs_debt_list($account) {
-  $result = array();
-  $res = array();
+function fbs_debt_list($account, $reset = FALSE) {
+  $results = &drupal_static(__FUNCTION__, array());
 
-  try {
-    $res = fbs_service()->Payment->getFees(fbs_service()->agencyId, fbs_patron_id($account), FALSE, FALSE);
-  }
-  catch (Exception $e) {
-    watchdog_exception('fbs', $e);
-  }
-
-  $ids = array();
-  foreach ($res as $fee) {
-    // The entity is only fetched in this case.
-    if (!empty($fee->materials) && count($fee->materials) == 1) {
-      $material = reset($fee->materials);
-      $ids[$material->recordId] = $material->recordId;
+  if (empty($results) || $reset) {
+    // Check if ding_session_cache is available.
+    if (module_exists('ding_session_cache') && !$reset) {
+      $results = ding_session_cache_get('fbs', 'debt', array());
     }
-  }
-  $ids = ding_provider_build_entity_id($ids);
-
-  foreach ($res as $fee) {
-    $id = $fee->feeId;
-
-    $data = array(
-      'date' => $fee->creationDate,
-      'display_name' => $fee->reasonMessage,
-      // FBS only provides the amount left to be paid on each fee, so this
-      // will only reflect that.
-      'amount' => $fee->amount,
-      // And the only thing we can do here is set it to zero, as the original
-      // amount isn't available.
-      'amount_paid' => 0,
-      'invoice_number' => NULL,
-      'type' => $fee->reasonMessage,
-    );
-
-    if (!empty($fee->materials)) {
-      $record_ids = array();
-      $material_ids = array();
-      foreach ($fee->materials as $material) {
-        $material_ids[] = $material->materialItemNumber;
-        $record_ids[] = $material->recordId;
+
+    // If both static cache and session cache failed, try reloading information
+    // from FBS.
+    if (empty($results) || $reset) {
```

Comment:
Does this introduce a problem about distinguishing between an empty cache and an empty debt list?

If the user does not have any debts then the cached result will be an empty array. Here this will be interpreted as an empty cache and thus result in debts being fetched again.

One way to approach this would be to use `NULL` as the default value for empty caches and checking it using `isset()`.

Reply:
This may give problems, so changed default value to NULL and using is_null to check it.